File: ruina-common/src/models/combat_page.ts
import { Chapter, Ego, Floor, Localization, Range, Rarity } from "../enums";
import { Die } from "./die";

export interface CombatPage {
    id: string;
    ego: Ego;
    egoFloor: Floor;
    scriptId: string;
    rarity: Rarity;
    dice: Die[];
    chapter: Chapter;
    cost: number;
    range: Range;
    imagePath: string;
}

export interface DecoratedCombatPage extends CombatPage {
    locale: Localization;
    name: string;
    description: string;
    diceDescriptions: string[];
}

File: rust/ruina_common/src/game_objects/combat_page.rs
use crate::game_objects::common::Chapter;
use crate::game_objects::common::Collectability;
use crate::game_objects::common::Rarity;

#[derive(Debug, Clone, PartialEq, strum_macros::Display)]
pub enum DieType {
    Slash,
    Pierce,
    Blunt,
    Block,
    Evade,
    CSlash,
    CPierce,
    CBlunt,
    CBlock,
    CEvade,
}

#[derive(Clone, Debug)]
pub struct Die<'a> {
    pub min: u16,
    pub max: u16,
    pub die_type: DieType,
    pub script: Option<&'a str>,
    pub actionscript: Option<&'a str>,
    pub motion: &'a str,
    pub effect_res: Option<&'a str>,
}

#[derive(Debug, PartialEq, strum_macros::Display)]
pub enum CombatRange {
    Melee,
    Ranged,
    Special,
    OnPlay,
    MassIndividual,
    MassSummation,
}

#[derive(Debug)]
pub struct CombatPage<'a> {
    pub id: &'a str,
    pub artwork: Option<&'a str>,
    pub cost: u8,
    pub range: CombatRange,
    pub rarity: Rarity,
    pub dice: &'a [Die<'a>],
    pub keywords: &'a [&'a str],
    pub options: &'a [&'a str],
    pub script_id: Option<&'a str>,
    pub skin_change: Option<&'a str>,
    pub map_change: Option<&'a str>,
    pub chapter: Option<Chapter>,
    pub priority: Option<i16>,
    pub collectability: Collectability,
}
